# День второй DWthon - выйди за рамки стандартного мышления!

## Наша цель - изучить закон Парето на практике!


Именно сегодня увидишь на практике, что есть такой закон, когда можно делать меньше, но получать больше! Даже встать на голову 🤸. Звучит нереально? Есть только один способ узнать :)

Кстати, заметь, что в обратную сторону очень легко. Быть занятым - очень легко, бегаешь и бегаешь, а по сути - мало кому это нужно.  Попасть в "черную дыру" очень легко. Почему это так? Выбрать правильные (важные) задачи достаточно сложно. Первое, что нужно знать в этом уроке, - это то, что есть такой закон! Второе - поговорим, как машинное обучение может помочь находить более важные задания, тем самым сосредоточимся на том, что важно.

### ☝️ Прежде, чем мы начнем, важно (!)

Во время наших инициатив мы много говорим о развитии. В этот раз мы тоже решили попробовать измерить твое развитие. Для чего?
Прежде всего, чтобы обратить Твое внимание на очень важную вещь, о которой, к сожалению, большинство забывает. Я имею в виду «дельту развития». Можно ли как-то измерить, насколько продвинулось Твое развитие (хотя бы в ML) во время DWthon? :) Давай попробуем!

Каждый день DWthon - проводится специальный опрос. Пожалуйста, заполни его, прежде, чем приступить к обучению. В конце концов - это может быть приятным сюрпризом :) Это займет у Тебя **только** 2-3 минуты.


##  👉👉 **[Заполни анкету](https://form.typeform.com/to/bxg2Ntuy)** 👈👈 и измерь дельту своего развития 💪

### Внимание! 

Я надеюсь, Ты помнишь, что это нормально, что у Тебя возникают вопросы. Если что-то не получается - обязательно об этом спроси в чате. Сегодня у нас второй день, поэтому все вопросы нужно задавать здесь [dwthon_day2](https://bit.ly/2UXq0mf)

Просто, подробно и по-человечески объясни, в чем возникла проблема. Лучше всего выложить скрин со своим кодом или обнаруженной ошибкой и написать, какой задачи это касается :). Также постарайся помочь другим.

In [1]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/NMeO61OnvGE" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

### Шаг за шагом

Я также записал для Тебя "пошаговое" видео. В видео ниже Ты найдешь именно то, что в этой записной книжке, только со всеми объяснениями, чтобы облегчить Твою работу и понять проблемы и задачи :) Посмотри, нужна ли Тебе такая дополнительная поддержка.

In [ ]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/ouKO8wwFg5o" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Закон 80/20

Это в первую очередь **практично** (в рамках DataWorkshop мы занимаемся практическими вещами, то есть вещами, которые хорошо работают в жизни). Стоит понимать, что закон 80/20 универсальный и применим ко многим аспектам нашей жизни (включая также Твою 😱, да, да - представь себе).

Приведу несколько примеров, чтобы стимулировать воображение:
- 20% сотрудников приносят 80% результатов;
- 20% людей являются причиной 80% автомобильных аварий;
- 20% продуктов - это 80% от общей суммы счета (посмотри, что это для Тебя, например, более крупные покупки в выходные);
- 20% населения несет 80% социальных расходов, включая преступность, медицинские услуги и т. д .;
- 20% исправленных ошибок сокращают 80% ошибок (которые действительно возникают);
- 20% кода обеспечивает 80% бизнес-функций (в том числе Microsoft подтвердил это в 2002 году).

---- 
Можно сказать, что существует закон природы, когда «меньшее делает большее, и наоборот». Этот закон получил свое название: Закон Парето. Важно, чтобы после этого урока Тебе стало очевидно, что этот закон пронизывает наш мир и нас, во многих измерениях нашей повседневной жизни.

Приведу наглядный пример. Какой это может иметь эффект именно для Тебя?

Если 20% времени может создать 80% ценности. Получается, 20% - это один рабочий день из пяти (в рабочей неделе). Думаем дальше 🤔. Представь, работаешь один день в неделю и создаешь 80% ценности от своей деятельности 😱. Получается, если последующие 4 дня - вообще ничего не делаешь, то особо никто этого не заметит? Может стоит работать только один день в неделю? 😉 Да, конечно, все зависит от контекста, и это не всегда просто, иначе все бы этим воспользовались, верно? 

С другой стороны это довольно-таки часто применимо, и поиск важных задач - это сложно назвать ракетостроением 🚀. Скорее всего, в большинстве случаев речь идет о том, чтобы изменить свое мышление и использовать правильные инструменты.

## Начинаем!
Конец размышлений, приступим к делу :) Сначала загрузим библиотеки, которые нам нужны на сегодня. 

In [2]:
import pandas as pd
import helper as h

Нам нужны `pandas`. Также добавил `helper` с более продвинутым кодом, который пригодится позже. Тебе не нужно анализировать код в `helper`, но если хочешь - то можешь это сделать в конце, когда уже сделаешь задания минимум.

## Загрузим данные
Надеюсь помнишь, как это сделать? Но повторение - мать учения! Мы используем для этого функции `pd.read_hdf`.

In [3]:
df = pd.read_hdf("../input/data.h5")

print(df.shape)
df.sample(5)

(820906, 9)


order_id  customer_id  product_id  quantity  price_unit  price_total  \
348400     22256         1336        3358         5         195          975   
323           35           23         130         2         125          250   
415236     25909          344         365         1          95           95   
104064      7360         1376        1116         1         125          125   
655424     43229          858          69         9          55          495   

        country_id          order_date  is_canceled  
348400           0 2011-11-14 08:29:00        False  
323              0 2010-12-01 11:21:00        False  
415236           0 2009-12-06 12:31:00        False  
104064           0 2011-04-10 12:52:00        False  
655424           0 2010-08-26 09:34:00        False

## Группировка

Во-первых, давай узнаем о другой полезной функции Python, а именно о `groupby`. Это функция группировки.

Чем это может быть полезно?

Допустим, мы хотим посчитать стоимость заказа.
Один заказ за раз может содержать один или несколько видов продуктов (и разное количество штук).

Давай рассмотрим это на примере. Картинка говорит больше, чем 1000 слов ;)
![](../images/order_rub_123.jpg)

Я опишу изображение, используя текущий формат данных. Смотри на картинку и на табличку.

Предположим, у хлеба ID 1 (`product_id=1`), помидоры имеют ID 2 (`product_id=2`)
у картофеля ID = 3 (`product_id=3`).
 
| order_id | product_id | quantity | price_unit | price_total |
| --- | --- | --- |--- | --- |
| 123 | 1 | 1 | 4 | 4 |
| 123 | 2 | 5 | 2 | 10 |
| 123 | 3 | 10 | 1 | 10 |

В этом примере (для простоты) я не указал, кто его купил, т.е. `customer_id`, потому что в данном случае это купил один покупатель, т.е. `customer_id` будет иметь одинаковое значение.

Удели несколько минут тому, чтобы посмотреть, что изображено на картинке, а что в таблице. Затем сделай себе тест. Глядя на картинку, попробуй переписать ее в виде таблицы. Почему это важно? Потому что тогда начнешь лучше понимать, «что происходит» в таблице, и меньше следовать просто инструкциям :). Ведь моя задача научить Тебя лично ловить рыбу, а не просто вручить удочку. Поработай над этим заданием 3-5 минут.

## 😱 Отсутствующие значения

В реальном мире часто бывает так, что иногда (на самом деле почти всегда в исходных данных 😂) у нас есть пропущенные значения. Это может вызвать некоторые проблемы, и Тебе также нужно знать, как с этим справиться (но я намекну, что в соответствии с правилом 80/20, которое  уже знаешь, с этим можно справиться довольно быстро).

С другой стороны, мы начнем с проверки наличия отсутствующих значений в наших данных. Как мы это сделаем?

Мы выполним 2 шага: 

1. Мы используем функцию `.isnull()`, которая возвращает `True`, если у нас есть отсутствующее значение, или `False` в противном случае. 
Другими словами, когда мы запустим `df.isnull()`, она вернет новую таблицу того же размера, что и наша исходная (т.е. `820906 x 9`) и будет содержать только значение `True` и/или `False`).

2. Затем мы суммируем количество значений `True` - т.е. отсутствующие значения. Любопытно, что в Python `True` автоматически преобразуется в 1, а `False` - в 0, и поэтому мы можем суммировать здесь.

Давай начнем!

In [4]:
df.isnull().sum()

order_id       0
customer_id    0
product_id     0
quantity       0
price_unit     0
price_total    0
country_id     0
order_date     0
is_canceled    0
dtype: int64

Мы видим, что конкретно в этом случае отсутствующих значений нет 🥳. С другой стороны, это не потому, что мир совершенен, просто эти данные "кто-то причесал", чтобы облегчить Тебе задачу. Это было сделано в целях науки для начинающих. 


### ☝️ Помни, что в исходных данных обычно есть отсутствующие значения.
Не стоит этого бояться 😉 Для начала, отсутствующие данные можно заполнить любым **уникальным** значением (например, -1). 


### Стоимость заказа
Как видно из таблицы, у нас есть вся информация с картинки ☝️, кроме одной. Стоимость всего заказа. Чтобы получить эту информацию, нам нужно сложить три значения (т.е. 3 строчки) из последнего столбца `price_total`. В одном заказае было три вида продуктов и каждый из них имел итоговую цену (`price_total`).



## Задание 2.1
Кстати, помнишь, как проверить, сколько у нас заказов? Давай повторим еще раз. Напиши код ниже.

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть подсказку 👈 </summary>
<p>
    <h3>💡 Подсказка</h3>

Если не особо помнишь, то "одним глазом" можно подсмотреть как мы это сделали в первый день (вчерашний ноутбук - day1).

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть ответ 👈 </summary>
<p>
    
```python
df["product_id"].nunique()
```
Старайся заглядывать в эту секцию только после того, как самостоятельно сделаешь задачу, чтобы сверить свое решение с нашим. 😉
</p>
</details> 
</p>
</details>

In [8]:
### YOUR CODE HERE
df['order_id'].nunique()

43908

Если задание выполнено, нужно сделать следующие:
1. Сделать скриншот, на котором видно решение задания
2. Добавить этот скриншот на канале [dwthon_day2_done](https://bit.ly/3x6zlFq)
3. Получить бонус за хорошую работу 💪 и завтра будет еще одна порция.

Внимание! Если у Тебя возникла проблемка с задачей или чего-то не знаешь, не помнишь, не понимаешь -  то смело спрашивай! Это обучение :) Задавать вопросы по 2-му дню DWthon нужно на канале [dwthon_day2](https://bit.ly/2UXq0mf).

## Цена заказа

Как же нам ее посчитать? Давай порассуждаем.
Нам нужно «как-то» собрать все `price_total` в одну "корзинку" для каждого заказа. 
Затем, когда у нас есть эти значения в одной корзинке, нам останется только суммировать `price_total`.

Получается нужно сделать два шага:
- группировка `df.groupby` (сбор в один мешок);
- сумма (или другое действие) значений в этом пакете `.agg`.

Посмотрим, как это выглядит в коде.

In [9]:
df[ ["order_id", "price_total"] ].groupby("order_id").agg('sum')

price_total
order_id             
0               27812
1                4440
2               55742
3               14010
4                3570
...               ...
52200            7080
52201           29746
52202            1095
52204          158039
52205           47351

[43908 rows x 1 columns]

Результат - каждая строчка - это сумма каждого заказа (например, сумма заказа на картинке была 24 рубля, помнишь?).

Давай еще один пример.
Например, мы уже знаем, что `order_id=0` стоит **278,12 фунтов стерлингов** (посмотри на первую строчку выше) или `order_id=52205` стоит **473,51 фунтов стерлингов**. ☝️ Напомню, что мы храним  `price_total` (и `price_unit`) в пенсах.

Давай еще раз рассмотрим этот код, чтобы убедиться, что мы все понимаем на 100%:
`df[["order_id", "price_total"]].groupby ("order_id").agg("sum")`


Мы можем разбить код на три логические части:
1. `df [[" order_id "," price_total "]]` из всех столбцов мы выберем только 2 столбца, потому что они нам нужны во 2-3 шаге
2. `groupby ("order_id")` мы "собираем" заказ за заказом в одну корзинку по ключу `order_id` (т.е. количество "корзинок" равно уникальному количеству `order_id`), другими словами, каждая корзинка - это отдельно взятый заказ со своим `order_id`;
3. `agg("sum")` как только у нас заказы разделены на отдельные корзинки, теперь берем все продукты в корзинке и суммируем `price_total` , так вычисляем сумму, которая и является стоимостью заказа (или нашей "корзинки"). 

Это похоже как в магазине. Длинная очередь покупателей и у каждого своя корзинка. Кассир "набивает" на чек все то, что находится в корзинке покупателя и выдает цифру - "итого". Затем повторяется то же самое.



## Сортировка
Было бы хорошо отсортировать наши заказы по убыванию. Самый дорогой заказ будет первым, а самый дешевый - последним. Для этого используем функцию `.sort_values()`. Пробуем написать в коде.

In [ ]:
(
    df[ ["order_id", "price_total"] ]
     .groupby("order_id")
     .agg('sum')
     .sort_values(by="price_total", ascending=False)
     .reset_index()
)

Появилась новая строка - `.sort_values(by="price_total", ascending=False)` - эта функция сортирует заказы по их стоимости. Обрати внимание на параметр `by=price_total`. Мы явно указываем название столбца, значение которого будет использоваться для сортировки. В данном случае в столбце `price_total` находится общая стоимость заказа.

Идем дальше. Параметр со странным и сложным именем `ascending` указывает направление сортировки. По умолчанию `ascending=True`, сортировка выполняется в порядке возрастания. В нашем случае мы хотим наоборот сделать - самый дорогой заказ должен быть первый и самый дешевый - последний.

`.reset_index()` я делаю техническую вещь, "переворачивая" значение `order_id` из индекса в столбец с именем (` order_id`). Если это звучит для Тебя слишком загадочно, можешь для начала проигнорировать это, т.к. это  - всего лишь крошечная деталь, которая не так уж и важна сейчас.

Обрати внимание, что наша строка с кодом становится все длиннее и длиннее.

```
df[ ["order_id", "price_total"] ].groupby("order_id").agg('sum').sort_values(by="price_total", ascending=False).reset_index()
```

Читать и добавлять новую цепочку кода - неудобно. Поэтому, расписали ее на меньшие части и добавили круглые скобки. Получилось вот так.

```
(
    df[ ["order_id", "price_total"] ]
     .groupby("order_id")
     .agg('sum')
     .sort_values(by="price_total", ascending=False)
     .reset_index()
)
```

Благодаря этому, код становится более читабельным и появляется возможность добавлять дополнительные условия (можно продолжить;)).

## Задание 2.2

Теперь сделаем то же самое, только ключ к "корзинке" будет `customer_id`. Другими словами, мы хотим подсчитать, какой доход принес нам каждый клиент. Например, Маша заплатила 10 000 рублей за все время существования магазина, а Саша 5 000 рублей и т.д.

Твоя задача - изменить «что-то» в предыдущем коде, чтобы получить общий доход для каждого покупателя.

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть подсказку 👈 </summary>
<p>
    <h3>💡 Подсказка</h3>

Вместо `order_id` используешь `customer_id`. 

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть ответ 👈 </summary>
<p>
    
```python
(
    df[ ["customer_id", "price_total"] ]
     .groupby("customer_id")
     .agg('sum')
     .sort_values(by="price_total", ascending=False)
     .reset_index()
)
```
Старайся заглядывать в эту секцию только после того, как самостоятельно сделаешь задачу, чтобы сверить свое решение с нашим. 😉
</p>
</details> 
</p>
</details>

In [13]:
### YOUR CODE HERE
(df[["customer_id", "price_total"]]
 .groupby('customer_id')
 .agg('sum')
 .sort_values(by = 'price_total',
              ascending = False)
 .reset_index()
)

customer_id  price_total
0             448     61130939
1             907     53007160
2            2647     33695110
3             179     31039769
4              66     29774259
...           ...          ...
5874         4594          295
5875          180          250
5876         4286            0
5877         4784            0
5878         4785            0

[5879 rows x 2 columns]

Если задание выполнено, нужно сделать следующее:
1. Сделать скриншот, на котором видно решение задания
2. Добавить этот скриншот в чате [dwthon_day2_done](https://bit.ly/3x6zlFq)
3. Получить бонус за хорошую работу 💪 и завтра будет еще одна порция.

Внимание! Если у Тебя возникла проблемка с задачей или чего-то не знаешь, не помнишь, не понимаешь, то смело спрашивай! Это обучение :) Задавать вопросы по 2-му дню DWthon нужно на канале [dwthon_day2](https://bit.ly/2UXq0mf).

## Меньшинство «делает» большинство, и наоборот.

Я обещал показать Тебе, что меньшинство порождает большинство и наоборот. Мы очень близки к этому, нам просто нужно получить сумму индивидуальных доходов для клиентов (это задание 2.2). 

Далее - добавляя каждую последующую строку (заказ) вместе, например, после 10 строк, мы получим общую сумму заказов для первых десяти клиентов. Когда мы сложим все строки, мы получим всю выручку магазина.

Давай покажу на коде - так будет проще объяснить, что имею в виду.

У нас есть функция для этого `cumsum()`.

In [14]:
sorted_orders = (
    df[ ["customer_id", "price_total"] ]
     .groupby("customer_id")
     .agg('sum')
     .sort_values(by="price_total", ascending=False)
    .reset_index()
)

sorted_orders["price_cumsum"] = sorted_orders["price_total"].cumsum()

sorted_orders.head(10)

customer_id  price_total  price_cumsum
0          448     61130939      61130939
1          907     53007160     114138099
2         2647     33695110     147833209
3          179     31039769     178872978
4           66     29774259     208647237
5          401     25800870     234448107
6           36     20169252     254617359
7           15     17923300     272540659
8         1168     15474456     288015115
9          862     15267156     303282271

Появилась новая строчка `sorted_orders["price_cumsum"] = sorted_orders["price_total"].cumsum()`

Теперь давай проанализируем.

Больше всего денег потратил покупатель `customer_id = 448`. В сумме этот покупатель потратил **611 309,39 фунтов стерлингов** (звучит, как стоимость многокомнатной квартиры в Москве, Киеве или Минске, но у людей разные аппетиты к подаркам).

Вторым «самым богатым» покупателем был `customer_id=907`, и он потратил **530 071,60 фунтов стерлингов**.

Теперь добавив `£611 309.39` (`customer_id=448`) + `£530 071.60` (`customer_id=907`), мы получим сумму `£1 141 380.99`. Это второе значение в `price_cumsum` (то есть сумма первых двух клиентов - их расходы). В третьей строке будет сумма трех заказов и так далее.
`£611 309.39` + `£530 071.60` + `£336 951.10` = `£1 478 332.09`

Надеюсь, уже понимаешь, что такое значения в столбце `price_cumsum` (на русском языке - это *кумулятивная сумма*, на английском - это `Cumulative Sum` или для краткости `cumsum`).

### 🧠 Включим критическое мышление.
Tеперь мы легко можем найти клиентов, которые генерируют 80%. Только давай сделаем дополнительную проверку, так называемый `double-check`, чтобы убедиться, что все в порядке. Давайте проверим выручку магазина двумя способами:
- сумма `price_total`
- последняя строка в `price_cumsum`


Значения должны быть идентичными. Почему? Кстати, подумай. Но стоит заметить, что предполагается в теории, довольно часто не совпадает на практике, так что нужно проверять и смотреть внимательно 👀.

In [15]:
df["price_total"].sum()

1816145630

Функция `tail` является обратной функцией` head` и позволяет получить последнюю строку.

In [16]:
sorted_orders["price_cumsum"].tail(1).values[0]

1816145630

Ну хорошо, как видно - значения совпадают.

У нас есть `£18 161 456.30 `, это больше `£18 mln.`. Теперь возьмем 80% от этой суммы.

In [17]:
1816145630 * 0.8

1452916504.0

80% составляет `£14 529 165.04`, что превышает £14 миллионов. А теперь давай возьмем самых «богатых» клиентов, которые генерируют нам эту сумму. В связи с тем, что наши клиенты уже отсортированы - остался один шаг.

In [20]:
value_80prc = int(1816145630 * 0.8)
top_clients =sorted_orders[ sorted_orders["price_cumsum"] < value_80prc ]["customer_id"]

Появилась строчка
- `sorted_orders[ sorted_orders["price_cumsum"] < value_80prc ]`

Эта конструкция может показаться трудной для начала. 

Посмотри, как это выглядит схематично:
`df[ df["a"] < 80 ]`

Получается так. Есть таблица (`DataFrame`) с именем` df`. Мы хотим отфильтровать строки, используя нужное нам условие. Это означает, что мы хотим оставить в таблице только те строки, которые удовлетворяют условию: например `df["a"] < 80`. Это условие означает, что оставляем только те строки, которые имеют значение менее 80 в столбце `a`.

В нашем случае - это то же самое, только имена переменных немного длиннее. Присмотрись:

```
sorted_orders[ sorted_orders["price_cumsum"] < value_80prc ]
```

Мы хотим оставить только тех клиентов, которые вместе (как можно "скорее") создают 80% нашей выручки.


🤔 Теперь момент истины .... Давай проверим, сколько клиентов приносят 80% выручки.

In [21]:
len(top_clients)

1324

У нас есть **1324** клиента, которые приносят 80% нашего дохода (более **£14 миллионов**). Сколько у нас всего клиентов, помнишь? Ты всегда можешь проверить еще раз;)

In [22]:
df["customer_id"].nunique()

5879

У нас всего **5 879** клиентов. Теперь разделим одно на другое.

In [23]:
1324 / 5879

0.225208368770199

Другими словами, в этом случае ситуация такова, что ~ 23% клиентов приносят 80% доходов.

### 😯 Но почему именно 23%, а не 20%?

Это зависит от многих факторов. Самым важным в этом правиле является то, что меньшинство (т. е. 23%) составляет большинство (т. е. 77%). Да, довольно часто это на самом деле не выглядит буквально - 20%, но не привязывайся к числу 20. На практике это может варьироваться от 5% до 30% (и считается меньшинством). Хотя стоит признать, что довольно часто этот показатель, действительно, очень близок к 20% (также и в этом примере).


Еще одно задание на сегодня 💪.

## Задание 2.3

Используя правило 80/20, мы применили его для клиентов, теперь пора сделать то же самое для продуктов. Другими словами, мы хотим проверить, работает ли это правило и для продуктов.

Возникает вопрос: **приносит ли меньшинство продуктов (около 20%) большую часть доходов (80%)**?

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть подсказку 👈 </summary>
<p>
    <h3>💡 Подсказка</h3>

- В этом случае вместо `customer_id`, должен быть использован `product_id`.
- Хорошо подумай, и у Тебя все получится. Здесь, действительно, есть что "немного" изменить. Как вариант, сделай небольшой перерыв, возможно, отдохни как-нибудь иначе 🛀 или более активно 🏊🏼🚴🏇 и снова вернись к задаче. Ты обязательно это сделаешь!

<details>
    <summary style="background: #e6eaeb; padding: 4px 0; text-align: center; font-size: 20px; font-weight: 900;"> 👉 Жми тут (1 клик), чтобы увидеть ответ 👈 </summary>
<p>
    
```python
sorted_products = (
    df[ ["product_id", "price_total"] ]
     .groupby("product_id")
     .agg('sum')
     .sort_values(by="price_total", ascending=False)
     .reset_index()
)
sorted_products["price_cumsum"] = sorted_products["price_total"].cumsum()

value_80prc = int(df["price_total"].sum() * 0.8)
top_products = sorted_products[ sorted_products["price_cumsum"] < value_80prc ]["product_id"]


prod_80revenu_perc = len(top_products) / df["product_id"].nunique()

print(f"80% дохода внесли {prod_80revenu_perc*100:.1f}% продуктов 🤔")
```
Старайся заглядывать в эту секцию только после того, как самостоятельно сделаешь задачу, чтобы сверить свое решение с нашим. 😉
</p>
</details> 
</p>
</details>

In [33]:
## YOUR CODE HERE
sorted_orders = (
    df[ ["product_id", "price_total"] ]
     .groupby("product_id")
     .agg('sum')
     .sort_values(by="price_total", ascending=False)
    .reset_index()
)
sorted_orders["price_cumsum"] = sorted_orders["price_total"].cumsum()
value_80prc = int(1816145630 * 0.8)
top_products =sorted_orders[ sorted_orders["price_cumsum"] < value_80prc ]["product_id"]
#len(top_products)
#df["product_id"].nunique()
prod_80revenu_perc = len(top_products) / df["product_id"].nunique()
print(f"80% дохода внесли {prod_80revenu_perc*100:.1f}% продуктов ")

80% дохода внесли 23.9% продуктов 


Если задание выполнено, нужно сделать следующее:
1. Сделать скриншот, на котором видно решение задания
2. Добавить этот скриншот на канале [dwthon_day2_done](https://bit.ly/3x6zlFq)
3. Получить бонус за хорошую работу 💪 и завтра будет еще одна порция.

Внимание! Если у Тебя возникла проблемка с задачей или чего-то не знаешь, не помнишь, не понимаешь - смело спрашивай! Это обучение :) Задавать вопросы по 2-му дню DWthon нужно в чате [dwthon_day2](https://bit.ly/2UXq0mf).

--- 
## Дополнительные задачи


Необязательно их делать, но если Ты чувствуешь «голод» и хочешь еще, то вперед 👍.

Время жонглировать данными и идеями 🤹.

### 🧠 Включим критическое мышление. 
Прямо сейчас мы думаем только о доходах. Однако стоит рассмотреть и тему расходов. Потому что это - неизменная вещь в любом бизнесе.

Каждый возврат - это не просто ноль - это, действительно, потеря. Почему?

Это связано с несколькими причинами:
- затраты на привлечение клиентов (например, стоимость рекламы);
- затраты на обработку заказа;
- затраты на проведение возврата (почта и т.д.);
- стоимость транзакции также платная (оплата через платежный шлюз, обычно один или даже несколько процентов перечисляются третьим лицам и не доходят до владельца бизнеса).

#### Возвраты, доходы и расходы
Давай сделаем график, чтобы в разных измерениях начать лучше чувствовать "природу" возвратов.

In [ ]:
h.plot_canceled(df)

In [ ]:
2615  / 3878

- Синий столбик - это доходы
- Оранжевый - это возвраты

У нас есть как минимум три измерения:
- клиенты (`customer`)
- заказы (`orders`)
- продукты (`products`)

Давай подумаем в этих трех измерениях:

### Клиенты
**5879** vs **2446**. 
Похоже, что по крайней мере 2446 клиентов (~41%) сделали возврат хотя бы один раз . Похоже, доходность составляет около.

### Заказы
**43 908** vs **7 290**. То есть возвращается только ~17% заказов.

## 😇 Включим творческое мышление
Теперь посмотри на измерения показателей и попробуй сделать собственные выводы. Что это значит для магазина?

--- 
## 🧠 Включим критическое мышление

Выше мы посчитали, сколько заплатил каждый клиент. Кстати, этот показатель называется `CLTV` или `Customer Lifetime Value` и означает, сколько покупатель будет тратить в магазине все время. Это важный параметр, который нужно уметь (точно) рассчитать. Это совсем непросто, но ML может здесь помочь.

Кстати заметь, что информация, сколько клиент уже потратил - не всегда синоним `CLTV`. Потому что у нас нет гарантии, что мы имеем историю с самого начала (возможно, тогда был другой магазин, другая база и многое другое), и мы не знаем, является ли это "концом" (последней покупкой этого клиента в этом магазине). Возможно, Маша только начинает и потратит еще в 10 раз больше, чем раньше. Очень сложно сказать, когда конкретный клиент не вернется ... Но стоит попробовать :).

## 😇 Включим творческое мышление

Подумай, как мы можем определить, что покупатель «никогда не вернется». Как к этому подойти, кроме того, что мы звоним и спрашиваем напрямую (что тоже малоэффективно, потому что покупатель всегда может передумать - всякое бывает), да и сама процедура неудобная (зачем беспокоить клиентов) и дорогая (сколько стоит 📞 обзвонить тысячи людей и провести с ними хотя бы 10-15 минут)?

Давай снова зададимся вопросом: **как мы можем определить, что покупатель совершил последнюю покупку в «магазине»?**. Почему так важно ответить на этот вопрос?

Мы хотим знать это, потому что тогда мы знаем, сколько мы можем «инвестировать» в клиента, потратив на его привлечение.

В качестве альтернативы стоит подумать, можем ли мы активировать клиента, который потенциально «забыл» о нас, и как это сделать с умом?


## 🤝🗣️ Совместное обсуждение и обмен опытом

Поделись, пожалуйста, своими мыслями с другими. Зачем? Благодаря этому могут родиться еще более качественные (совместные) идеи. Это когда 1+1 = 11 😉

Напиши в чатике [dwthon_day2_ideas](https://bit.ly/3zCetr4), что у Тебя получилось выдумать. Поучаствуй в обсуждении, и увидишь на примере - как легко генерировать множество идей в сотрудничестве с другими. Это очень полезный навык, который скорее всего Тебе пригодится в современном мире.

## 🎂 Десерт

Я подготовил для Тебя немного более сложное задание, чтобы Ты смог построить графики и посмотреть своими глазами 👀 как меньшее делает большее и наоборот, в трех измерениях: клиенты, товары, заказы, а также можешь переключаться между дополнительными двумя: выручка, прибыль;

Покажу на примере. Начнем с клиентов.

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_80',
    'main_feat': 'customer_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

- На оси `X` клиенты в % - возможные значения: от 0 до 100%
- На оси `Y` доходы в % - возможные значения: от 0 до 100%


Зеленая вертикальная линия показывает места, где мы получаем 80%. 

Давай вместе проанализируем график.
Зеленая (вертикальная линия) пересекается на оси X там, где число 22,5 и на оси Y там, где значение 80. Что означает, что 22,5% клиентов (наиболее "богатых") создают 80% доходов.

Обрати внимание, как быстро синяя линия стремится "вверх" и потом сбавляет обороты. В этом и заключается закон Парето, когда меньшее количество клиентов дает большее количество доходов.

Давай еще немного комментариев по поводу кода.
```
args = {
    'df': df,
    'sel_x_prct': 'auto_80',
    'main_feat': 'customer_id', 
    'price_feat': 'price_total', 
}
```

Здесь видны 4 параметра:
- `df` - это наш таблица (`DataFrame`)
- `sel_x_prct` - параметр, где делаем зеленую черту, `auto_80` значит, что автоматически находим значение на оси `X`, которое соответствует 80% на оси `Y`, например, можешь написать `'sel_x_prct': 'auto_50'`

Тогда найдем сколько процентов покупателей создают 50% прибыли.

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_50',
    'main_feat': 'customer_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

Видишь, 4,1% клиентов генерируют 50% доходов 😱. Так и хочется спросить - а что делают остальные 95,9%?

Третий параметр был: `'main_feat': 'customer_id',`. Здесь можно переключаться между:
- клиентами: `customer_id` 
- продуктами: `product_id` 
- заказами: `order_id` 

Давай сразу применим и посмотрим на примере - сколько процентов продуктов генерируют 80% доходов ( `'main_feat': 'product_id'` )

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_80',
    'main_feat': 'product_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

925 продуктов из 3878 продуктов (т.е. ~23.9%) генерируют 80% доходов.

Иными словами, смотри на этот график, мы можем предположить, что если мы удалим из нашего ассортимента около 3000 продуктов (3878 - 925 = 2953), то доход уменьшится только на 20%! У нас будет на 76% меньше головной боли, что касается ассортимента, и только на 20% меньше дохода. А что если мы нашу энергию и внимание перенесем туда, чтобы добавить еще, например, 100-200 продуктов, которые похожи на те 925 продуктов, чтобы доход был такой же (или даже больше), а работы меньше?

Именно к этому мы и стремимся. Делать то, что имеет ценность для других людей и это решает какие-то их проблемы (потребности).


Ну хорошо, а давай проверим сколько продуктов создают 50% доходов. Для этого нужно поменять `'sel_x_prct': 'auto_50',`. Ну что, смотрим?

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_50',
    'main_feat': 'product_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

Получается, что 250 продуктов (6,4% из общего ассортимента) приносит 50% дохода! 😱 Если мы каким-то образом найдем еще таких 250 продуктов, то получится, что доход останется таким же, но продуктов будет меньше. Ведь содержание каждого продукта - это затраты.

Давай еще посмотрим на заказы. Работает ли и в этом случае закон Парето?

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_80',
    'main_feat': 'order_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

Этот случай менее выраженный - должно произойти около 39% заказов, чтобы мы смогли получить стоимость в 80%. Но все-таки,
по-прежнему 39% - это меньшинство ;). Даже здесь, обслужив достойно только 39%, получим 80% дохода.

Давай проверим для 50% доходов.

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_50',
    'main_feat': 'order_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

Получается, что 12% заказов генерируют половину дохода. Хотелось бы разобраться с "природой" более выгодных заказов и имеем ли мы возможность привлечь еще таких 12% заказов, чтобы реализовывая 1/4  заказов, получить 100% дохода? Звучит заманчиво.



### Что думаешь по этому поводу?

In [ ]:
## Напиши свои мысли и поделись ими в Slack :)
...


Кстати, мы до этого анализировали доходы, т.е. игнорировали возвраты. Давай более детальнее посмотрим на "выручку", т.е. учтем возвраты как ноль (хотя на самом деле возвраты - это обычно всегда минус для бизнеса, т.к. работа уже была проделана).

Для этого нужно поменять четвертый параметр в `args`:
- было: `'price_feat': 'price_total',` 
- стало: `'price_feat': 'profit_total',` 

Вот пример для клиентов.

Только для начала посчитаем `profit_total`. Идея простая, если возврата не было, то оставляем значение `price_total`, в другом случае будет 0. 

In [ ]:
df["profit_total"] = df["price_total"] * ~df["is_canceled"]

Теперь можно строить график прибыли для клиентов.

In [ ]:
args = {
    'df': df,
    'sel_x_prct': 'auto_80',
    'main_feat': 'customer_id', 
    'price_feat': 'profit_total', 
}

h.plot_pareto(**args)

"Картинка" похожая, как была с доходами, но все же немного отличается.
- Доходы - 22.5% это 1324 клиентов
- Прибыль - 23.0% это 1351 клиентов

Теперь можешь дальше проверить и построить похожие графики. ☝️ Помни только, что для графиков с прибылью нужно использовать `'price_feat': 'profit_total'`.

In [ ]:
args = {
    'df': df[df.is_canceled],
    'sel_x_prct': 'auto_80',
    'main_feat': 'order_id', 
    'price_feat': 'price_total', 
}

h.plot_pareto(**args)

## Отзыв и дельта Твоего прогресса

Дай мне, пожалуйста, знать, что Тебе понравилось, было ли трудно (и где именно) и зафиксируй также для себя - чему получилось научиться на этом уроке.

### Тут 👉👉 [анкета](https://bit.ly/2TI3FbM) 👈👈
Твой отзыв позволит улучшить материал!

## До завтра! 🙌